File: 0120-triangle/0120-triangle.py
class Solution(object):
    def minimumTotal(self, triangle):
        m=len(triangle)
        def lol(i,j):
            if i == 0 and j == 0:
                return triangle[0][0]
            if j < 0 or j > i:
                return float('inf') 
            return triangle[i][j] + min(lol(i-1, j-1), lol(i-1, j))
        last = m - 1
        return min(lol(last, j) for j in range(last+1))

        

File: 0120-triangle/0120-triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int n= triangle.size();
        int[][]dp=new int[n][n];
        for(int i=0;i<n;i++) Arrays.fill(dp[i],Integer.MIN_VALUE);
        return solve(0,0,triangle,dp);
    }
    private int solve(int i,int j, List<List<Integer>> triangle,int[][]dp){
        if(i==triangle.size()-1) return triangle.get(i).get(j);
        if(dp[i][j]!=Integer.MIN_VALUE) return dp[i][j];
        return dp[i][j]=triangle.get(i).get(j)+Math.min(solve(i+1,j,triangle,dp),solve(i+1,j+1,triangle,dp));
    }
}